### PR TITLE
Added support for searchengines listing in :open dialogue. 

### PR DESCRIPTION
**Pull request by philipcs**
Primarily, this change adds support for showing search engines in the :open dialogue, since I find myself using them a lot more than I use any quickmarks or history while at work. Since I want mine to be quick to type I have a tendency to use too short prefixes for these, which results in me forgetting them. Being reminded what search engines are available helps me, since I often use the "Search engine" feature as a shortcut into systems like internal wikis or case management systems.

I have also added options to select which categories of lists should be shown in the :open dialogue (history, bookmarks, quickmarks and/or search engines - see url.open_categories_shown).

I'm less confident that my additions to the settings.asciidocs are correct. I have tried to imitate the other entries as best as I can. I have not updated the tests, nor included any new ones for this.

Is this something you could consider adding to the codebase?

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4004)

### CODE REVIEW

**Comment by The-Compiler** [comment]
This definitely looks like something useful, and you're not the first to ask for it! Also see #836 and #32.

cc @rcorre who might want to take a quick look, I'll probably follow up with a review next week (or this week, depending on your timezone :wink:)

> I'm less confident that my additions to the settings.asciidocs are correct. I have tried to imitate the other entries as best as I can. I have not updated the tests, nor included any new ones for this.

That file is autogenerated - see the comment at the beginning of the file.

File: qutebrowser/config/configdata.yml [hunk 1]
```diff
@@ -1565,6 +1565,19 @@ url.open_base_url:
...
+  type:
+    name: List
+    valtype: String
+    none_ok: true
+  default:
+    - searchengines
+    - quickmarks
+    - bookmarks
+    - history
+  desc: Which categories should be shown in the :open dialogue.
```
Comment (mschilli87) [inline qutebrowser/config/configdata.yml]: I'd suggest `Which categories to show in the :open dialogue.` for consistency (see https://github.com/qutebrowser/qutebrowser/commit/01b29cd64051d55c73102e8c6062fb743243bbd0).

**Comment by philipcs** [comment]
@jgkamat I like that idea. Since I was uncertain if the code would be accepted, I thought I would keep it simple, but I added your suggestion to my branch.

File: qutebrowser/config/configdata.yml [hunk 2]
```diff
@@ -1565,6 +1565,19 @@ url.open_base_url:
...
+  type:
+    name: List
+    valtype: String
+    none_ok: true
+  default:
+    - searchengines
+    - quickmarks
+    - bookmarks
+    - history
+  desc: Which categories to show in the :open dialogue. The order of this list is used for ordering the categories.
```
[the remark this answers is not in the document]
Comment (philipcs) [reply qutebrowser/config/configdata.yml]: How about 
```
desc: Which categories to show in the :open dialogue. The order of this list is used for ordering the categories. Valid options are zero or more of "history", "quickmarks", "bookmarks" and "searchengines". If you select none, the :open suggestion dialogue will not be shown.
```

**Comment by philipcs** [comment]
I'll take a look at the tests when I get some time today/tomorrow, but I'm afraid I have never used any of your testing frameworks before.

**Comment by rcorre** [comment]
Neither had I ;)

https://www.qutebrowser.org/doc/contributing.html may be of use. I believe you'll want to work with `tests/unit/completion/test_models.py`.

File: qutebrowser/config/configdata.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (The-Compiler) [reply qutebrowser/config/configdata.yml]: The valid values should really be given in `configdata.yml` instead (probably by making this a `FlagList` too). I'll comment on the rest tomorrow, didn't get around to it today :wink:

**Comment by philipcs** [comment]
I'm working on the tests still, I think it may take a few days due to the day job. 
One question occurred to me while setting up tests, however. What do you think of showing the DEFAULT search engine in the listing? I am leaning towards hiding it, but I can't really decide this would be unexpected to the end users or not. I could make it configurable, but I'm a bit worried about polluting the settings page with too many options.

**Comment by jgkamat** [comment]
philipcs writes:

> One question occurred to me while setting up tests, however. What do you
> think of showing the DEFAULT search engine in the listing? I am leaning
> towards hiding it, but I can't really decide this would be unexpected to the
> end users or not. I could make it configurable, but I'm a bit worried about
> polluting the settings page with too many options.

I would personally vote for omitting DEFAULT without an option to change it, I
think it's redundant to not selecting anything.

**Comment by philipcs** [comment]
Had to delay a bit due to work. The changes mentioned above have been committed locally. I'll put aside some time to learn pytest this weekend.

**Comment by The-Compiler** [comment]
I'm tuning out of qutebrowser stuff because of my exams now, so take your time! I'll get back to stuff in September.

**Comment by philipcs** [comment]
I got a bit distracted by vacation and a problem at work. Apologies for
that.

I was planning to set aside some time this weekend and hopefully finish
this.

So the issue is not dead yet ;)

**Comment by philipcs** [comment]
I am still struggling with making the tests work out for me. I am fairly certain it's a PEBKAC, but I was hoping you could give me a hint as to which direction I  need to go.

First of all: where do I set the config settings. I thought I would do it by changing the `configdata_stub`, but I found that it was overridden by the `configdata.yml` config file (or rather that it did not use the `default` values).

I tried to do it by providing a fixture like this, but I suspect I am doing it wrong.
```
@pytest.fixture
def searchengines(searchengines_manager_stub):
    """Pre-populate the searchengine-manager stub with some searchengines."""
    searchengines_manager_stub = collections.OrderedDict([
        ('DEFAULT', 'https://duckduckgo.com/?q={}'),
        ('google', 'https://google.com/?q={}'),
        ('another', 'https://google.com/test?q={}'),
        ('another2', 'https://google.com/test2?q={}'),
    ])
    return searchengines_manager_stub
```
But as I was working on that I came to the conclusion that the only examples I have seen in the code were using stubs (which again were using *-manager - like bookmark-manager). I have not been using that in the actual code for searchengines (and you did not comment on needing it earlier) so I suspect it might be overkill to implement it. Should I be setting the `config.val.url.searchengines` setting instead of providing the stub?

**Comment by toofar** [comment]
I'm not sure I'm parsing your question right withoutout all the context but I think you should be configuring the config_stub stub in your tests like here https://github.com/qutebrowser/qutebrowser/blob/0a9806daf3fb93f1525539eb82455305fd79d099/tests/unit/completion/test_models.py#L519

**Comment by The-Compiler** [comment]
Sorry for the long delay! I merged this and pushed a couple of fixes on top:

- 81e5173c9cece5ae8eb9f97e9133b9d28ff58bc4 makes sure the search engines are sorted (dict iteration order is undefined on Python 3.5), simplifies the dict to a list comprehension, and uses `k != 'DEFAUlT'` so e.g. `'DEF'` isn't filtered as well.
- 4b13609553083abd7d9963d6e86a0fc12e3a3c76 reverts some probably unintended change and makes pylint/flake8/me happy with the coding style :wink:

Thanks!

**Pull request merged**